### PR TITLE
New replay cache implementation

### PR DESCRIPTION
**Pull request by greghudson**
I also want to simplify the internal rcache interface as a final step, but this is already a big chunk to review.

I haven't seen code from Roland, so this PR contains my implementation.  I have some notes about performance testing at https://k5wiki.kerberos.org/wiki/Projects/Replay_cache_improvements#Hash-based_file_format .  I wound up using 128-bit records (so no aggressive trimming of the tag or timestamp), seeded siphash-2-4 over the tag to allay my concerns about hash table performance attacks, and whole-file locks.  Nothing is set in stone, though.

Tagging @nicowilliams since he may have comments on the details of the format.  I didn't find a github account for Roland.

### CODE REVIEW

**Comment by nicowilliams** [comment]
Roland is @elric1 (paging @elric1).

**Comment by nicowilliams** [comment]
Note that rcache file naming may be tricky where services are keyed by AD sharing a machine account for several services...

**Comment by greghudson** [comment]
I guess we could drop the server name and just use the euid?  Hopefully all server processes which could accept the same ticket are running at the same euid; otherwise it's hard to share a replay cache between them.

**Comment by nicowilliams** [comment]
Byte range file locking might be more appropriate than whole file locking, but that's got all the pitfalls of POSIX file locking...

**Comment by nicowilliams** [comment]
Maybe hash the key?

**Comment by greghudson** [comment]
My observations were that (1) performance in my tests seemed to be dominated by system call overhead rather than concurrency limitations, and byte-range locking takes more system calls, and (2) we can migrate from whole-file locking to byte-range locking with no compatibility issues, since whole-file locks contend with every byte-range lock in the same file.

**Comment by nicowilliams** [comment]
Fair enough.

**Comment by nicowilliams** [comment]
We do have to get the format right before we start having to interop with this one forever.

I'm still keen on being able to use atomic CAS through `mmap()` as a way to increase concurrency and decrease system call overhead.  Using `mmap()` will not always be feasible -- it will depend on the OS and filesystem supporting this and having a unified block cache.

One possibility is to have 1 bit from the 128-bit entry size dedicated to being a "busy" bit, and then use it to make a spin lock -- but I'm not keen on spin locks.  We could make that safer by having a header with N slots for active threads to write their PID/TID into, and this could be used for recovery from timeout, but again, that's a bit complex.

The motivation for any complexity is performance.  I... don't have numbers on what performance we need from the rcache (we currently disable it).  We should do try to characterize required performance before we finalize this.  I'll try to get some numbers.

**Comment by nicowilliams** [comment]
Regarding sharing of keys... The simplest thing to do is to let the user specify a non-default rcache path.  That's not the safest thing to do -- the user may not understand how AD machine accounts work.  There's a trade-off here.

Multiple services running as different users can still share an rcache, but it does require that the admin setup some group or file ACL for it.  Failure to do so with a default rcache name based on a partial hash of the key would lead to failure, which would be detected.  OTOH, using the service name or EUID would cause shared-key situations that don't share an rcache to go undetected.

The perfectionist in me wants to use a partial hash of the key.  The pragmatic in me wants to use the EUID.  I'm really not sure what the best thing to do here is.

**Comment by greghudson** [comment]
I will note that when we decrypt a ticket in krb5_rd_req(), we iterate over the keytab (perhaps filtered by etype and by the calling application's server name) and pick the first key that works.  We use the name of that keytab entry to pick the rcache name.  We don't pay any attention to the name presented in the ticket, with the exception of when we're using the KDB as keytab (krb5kdc and kadmind).

There could still be a problem if the same key appears in multiple keytab files with different names, of course.

**Comment by nicowilliams** [comment]
Right, that's the point.  Because of AD this multiple-services-sharing-one-key thing happens.  Not sharing an rcache among them is clearly wrong.

**Comment by nicowilliams** [comment]
BTW, Heimdal has hard principal aliasing support now.  I think that means effectively sharing keys between them...

**Comment by nicowilliams** [comment]
At today's krbdev meeting we discussed the perf. requirements / locking model, and the issue of how to name the rcache file.  I'll summarize our proposal for the latter here:

 - fail open -- it would be disastrous for MIT or Heimdal to start failing on upgrade due to inability to setup an rcache
 - but allow applications that use `gss_acquire_cred_from()` to require fail-closed, and optionally to specify an exact rcache path.

 - use `/var/krb5/rcache-v2/<principal_name>` if it exists (but do not create if missing)
 - use `/var/krb5/rcache-v2/<base32(truncate(4, SHA256(truncate(8, key))))>` if it exists (but do not create if missing)
 - use or create `/var/run/user/<EUID>/rcache-v2`
 - failing that, use or create `/run/user/<EUID>/rcache-v2`
 - failing that use a MEMORY rcache

**Comment by nicowilliams** [comment]
A version number needs to be included in the rcache file name/path.

If ever we add a new version, we'll have to search first for older versions, then newer, and if an older version is present, then we must use that.  This is to support mixed version deployments (some apps use version X of the Kerberos libraries, while others use version X+N).

**Comment by greghudson** [comment]
On the call, Nico suggested that we stop using replay caches for KRB-SAFE/KRB-PRIV/KRB-CRED messages, which we currently do when the auth context has the DO_TIME flag set (as is unfortunately the default).  That would probably be okay--kpasswd and kprop use DO_SEQUENCE rather than DO_TIME, and Heimdal just has a long-standing `/* XXX - check replay cache */` comment for the DO_TIME case.

However, here Nico suggests falling back to an in-memory replay cache if we can't create one.  If we create an in-memory rcache facility, we could use it for DO_TIME auth contexts.  At least then if a subkey is used, replays and reflections would be detected, as the scope of potential replays would be limited to an auth_context in one process.
 
We have an in-memory lookaside cache in kdc/replay.c which isn't much code (now that it uses k5-hashtab.h) and could easily be simplified down for this purpose.  So my immediate plan is to file a new PR to:

1. Add an in-memory rcache facility to libkrb5, with a simple API separate from the krb5_rcache API.
2. Make KRB-SAFE/KRB-PRIV/KRB-CRED use this facility for DO_TIME, borrowing from this PR's second commit to generate the unique tags.

Then, for this PR, we will only have to consider the use case of krb5_rd_req(), and we will have the in-memory rcache facility available to fall back to when we cannot find or create a filesystem rcache.

**Comment by greghudson** [comment]
Rebased on top of the memory ccache work, but I haven't updated the naming yet.

> use /var/krb5/rcache-v2/<principal_name> if it exists (but do not create if missing)
> use /var/krb5/rcache-v2/<base32(truncate(4, SHA256(truncate(8, key))))> if it exists (but do not create if missing)

These are intended to allow the system to provision replay caches if a principal is used across multiple euids, with group write access for a group containing all of the euids.  We don't expect this to be common.  This will come at the cost of a failed open() for most uses; we could remember the failure for a given replay cache handle.

I would prefer to do just one of these.  On the call I think we decide that we think Windows only shares keys between different SPNs for RC4, so the hashed key form (which would be harder to use) may not be necessary.

Nico talked about using the full principal name rather than the first component.  I guess the theory is that the same system might have different euid use sets for different host or HTTP principals or whatever.  That seems kind of exotic, and we have the existing API krb5_get_server_rcache() which accepts just the first component.  That's not a true blocker because the API is mostly only used internally, but it's easier if we stick with the first component.

> use or create /var/run/user/<EUID>/rcache-v2
> failing that, use or create /run/user/<EUID>/rcache-v2

If the first of these doesn't work, is there value in trying the second?  Do we think there are systems with only /run/user?  I'm having some trouble chasing down any attempt at standardizing these directories; I had considered using an XDG_ environment variable to find them, but only XDG_RUNTIME_DIR exists and that's the one we'd prefer less.

> A version number needs to be included in the rcache file name/path.

I would prefer to put this in a filename extension (like `.rcache2`) instead of a directory.

Other questions I will need to answer:

* Do we still obey KRB5RCACHEDIR, and if so, what filename do we use there?  (If not, we need another solution for the test suite.)

* Do these pathnames interact with configure options in any way?

* What do we use on Windows?

**Comment by greghudson** [comment]
A candidate plan after another round of voice discussion on (I think) the April 30 call:

* Introduce the new replay cache type as "file2", where the residual is a pathname to the file (similar to the FILE ccache or keytab type).  file2 fails closed, meaning it results in errors if the specified pathname can't be used.

* The "dfl" type is rewritten as a shim.  The shim checks if /var/run/user/EUID is a directory.  If yes, it devolves to file2:/var/run/user/EUID/krb5.rcache2; if no, it uses a process-wide memrcache.  [Autoconf: we use the expansion of runstatedir in place of /var/run.  Windows: not sure yet.]

* krb5_get_server_rcache() ignores the piece argument and just resolves the default rcache.  "Resolving the default rcache" has the following hooks from highest preference to lowest:
** If KRB5RCACHENAME is set in the environment, resolve it.
** If KRB5RCACHETYPE is set, resolve it with no residual.
** If KRB5RCACHEDIR is set, resolve file2:KRB5RCACHEDIR/krb5.rcache2.
** If [libdefaults] default_rcache_name is set, expand path tokens in it and resolve that.
** Otherwise resolve "dfl:".

Relative to the current code: KRB5RCACHENAME is new, sort of--it's consulted in the code in krb5_rc_default_name() and krb5_rc_default(), which are unused non-API functions.  KRB5RCACHETYPE is currently set to "none" by some applications to disable the rcache, and that will continue to work.  KRB5RCACHETYPE=dfl will also continue to work, although there's not much reason for anyone to set it.  KRB5RCACHEDIR is currently used inside the dfl type; making it use the file2 type makes it continue to work in spirit.  [libdefaults] default_rcache_name is new.

**Comment by greghudson** [comment]
I realized that on systems with a /var/run/user/EUID, /var/run is typically a symbolic link to /run, so it's going to be the same as /run/user/EUID, which (like everything else in /run) gets cleared out on boot.  There is no per-user analog to /var/tmp -- a directory on local storage which is not cleared out on reboot.

Option 1 is to use it anyway (perhaps via $XDG_RUNTIME_DIR).  As we're not willing to pay the performance penalty of using fsync(), replay records aren't guaranteed to survive reboot, so we could also decide that it's okay to just lose them all across a reboot.

Option 2 is to continue using /var/tmp, and try to make that safe.  open() with O_CREAT and without O_EXCL will follow symbolic links, so I think we need O_NOFOLLOW (from POSIX.1-2008 and some systems before that) to use /var/tmp safely.  And of course there would be a trivial denial attack by any other uid.

**Comment by frozencemetery** [comment]
Additional consideration: {,/var}/run is more likely to be a tmpfs -
despite the names, /tmp and /var/tmp aren't usually by default.

**Comment by greghudson** [comment]
Updated, using /run for now.  Losing all replay cache records across a reboot is unfortunate, but modern systems don't seem to provide a good filesystem location to ignore that (/var/tmp being a bad location due to being shared across users).  Downstream integrations or system administrators have a lot of hooks to change the replay cache location if they want.

**Comment by greghudson** [comment]
Back to /var/tmp, with the euid in the filename.  I made the code unconditionally use O_NOFOLLOW after verifying that my Solaris build worker has it.

I preserved the separation of dfl and file2, and the addition of KRB5RCACHENAME and [libdefaults] default_rcache_name.  The idea is that using file2 will work if the pathname is a symlink or if the file isn't owned by the euid (which would have to be the case for a pre-provisioned rcache shared across euids), while the default type has the safety checks needed to use a shared temporary directory.

**Comment by frozencemetery** [comment]
Right.  We've hacked around this for the /run/user/0 case [downstream](https://github.com/frozencemetery/krb5/commit/75ba8f42c0e9426af80c71aaaa490cc6262e259c) though I still need to investigate whether this actually helps anything, and it'd need to be decoupled from the SELinux patch.

File: doc/basic/rcache_def.rst [hunk 1]
```diff
@@ -71,21 +71,30 @@ are in lowercase.  The following types are defined:
 
 #. **dfl** is the default type if no environment variables or
    configuration specifies a different type.  It stores replay data in
-   a file, occasionally rewriting it-to purge old, expired entries.
+   a file2 replay rache with a filename based on the effective uid.
```
Comment (frozencemetery) [inline doc/basic/rcache_def.rst]: "replay rache" should be either "replay cache" or "rcache".

**Comment by greghudson** [comment]
> rc_base.h can go while you're here, if you're so inclined.

Out of scope, I think, since it's still included from rc_conv.c and rc_base.c and was empty before this PR.  If I get around to simplifying the rcache interface (getting rid of most of the methods except store, etc.) then it can be removed with that work.

**Comment by frozencemetery** [comment]
Sounds good.

File: src/lib/krb5/rcache/t_rcfile2.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Usage:
```
[the remark this answers is not in the document]
Comment (greghudson) [reply src/lib/krb5/rcache/t_rcfile2.c]: It could, but it's a test program run only by `make check`, so I think a comment is fine.

**Pull request merged**